core/src/test/java/kafka/testkit/junit/ClusterForEach.java

```diff
@@ -0,0 +1,246 @@
+package kafka.testkit.junit;
+
+import kafka.api.IntegrationTestHarness;
+import kafka.network.SocketServer;
+import kafka.server.Kip500Broker;
+import kafka.server.Kip500Controller;
+import kafka.server.LegacyBroker;
+import kafka.testkit.ClusterHarness;
+import kafka.testkit.KafkaClusterTestKit;
+import kafka.testkit.TestKitNodes;
+import org.apache.kafka.common.network.ListenerName;
+import org.apache.kafka.common.security.auth.SecurityProtocol;
+import org.apache.kafka.metadata.BrokerState;
+import org.junit.jupiter.api.extension.AfterTestExecutionCallback;
+import org.junit.jupiter.api.extension.BeforeTestExecutionCallback;
+import org.junit.jupiter.api.extension.Extension;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.TestTemplateInvocationContext;
+import org.junit.jupiter.api.extension.TestTemplateInvocationContextProvider;
+import org.junit.platform.commons.util.ReflectionUtils;
+import scala.Option;
+import scala.jdk.javaapi.CollectionConverters;
+
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Properties;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Template that creates two test invocations: one for legacy cluster and one for quorum clusters.
+ *
+ * In addition to providing the BeforeEach and AfterEach extensions for these generated invocations, this provider
+ * also binds two parameters which can be injected into test methods or class constructors:
+ *
+ * <ul>
+ *     <li>IntegrationTestHelper: Utility class of common network functions needed by integration tests</li>
+ *     <li>ClusterHarness: Interface that exposes aspects of the underlying cluster</li>
+ * </ul>
+ */
+public class ClusterForEach implements TestTemplateInvocationContextProvider {
```

Comment:
Rather than hard-coding this as a reference to either the kip-500/zookeeper clusters, I wonder if it makes sense to see this as a parameterization of `KafkaConfig`. We could still let `provideTestTemplateInvocationContexts` determine the right test harness based on the presence of `process.roles` or `zookeeper.connect`. But then it is general enough to handle other cases. For example, we might want an integration test which verifies behavior of the KIP-500 controller if it is started either as "broker" or "broker,controller." Maybe it could look something like this:

```java
@ExtendWith(value = Array(classOf[ClusterForEach]))
@ClusterParameterizedTest(config=List(
  singletonMap("process.roles", "broker"),
  singletonMap("process.roles", "broker,controller"),
 )
```

Then we could define `ClusterParameterizedTest` at either the class or method level.